src/porepy/numerics/time_step_control.py

```diff
@@ -0,0 +1,273 @@
+            raise ValueError("Maximum number of iterations must be > 0")
+
+        if iter_optimal_range[0] > iter_optimal_range[1]:
+            s = "Lower optimal iteration range cannot be larger than"
+            s += " upper optimal iteration range."
+            raise ValueError(s)
+        elif iter_optimal_range[1] > iter_max:
+            s = "Upper optimal iteration range cannot be larger than"
+            s += " maximum number of iterations."
+            raise ValueError(s)
+
+        if iter_lowupp_factor[0] <= 1:
+            raise ValueError("Expected lower multiplication factor > 1")
+        elif iter_lowupp_factor[1] >= 1:
+            raise ValueError("Expected upper multiplication factor < 1")
+
+        if recomp_factor >= 1:
+            raise ValueError("Expected recomputing multiplication factor < 1")
+
+        if recomp_max <= 0:
+            raise ValueError("Number of recomputing attempts must be > 0")
+
+        # Schedule, initial, and final times
+        self.schedule = schedule
+        self.time_init = schedule[0]
+        self.time_final = schedule[-1]
+
+        # Initial time step
+        self.dt_init = dt_init
+
+        # Minimum and maximum allowable time steps
+        self.dt_min, self.dt_max = dt_min_max
+
+        # Maximum amount of iterations
+        self.iter_max = iter_max
+
+        # Target iteration range
+        self.iter_low, self.iter_upp = iter_optimal_range
+
+        # Lower and upper multiplication factors
+        self.iter_low_factor, self.iter_upp_factor = iter_lowupp_factor
+
+        # Re-computation multiplication factor
+        self.recomp_factor = recomp_factor
+
+        # Number of permissible re-computation attempts
+        self.recomp_max = recomp_max
+
+        # Time
+        self.time = self.time_init
+
+        # Time step. Initially, equal to the initial time step
+        self.dt = self.dt_init
+
+        # Private attributes
+        # Number of times the solution has been recomputed
+        self._recomp_num = 0
+
+        # Index of the next scheduled time
+        self._scheduled_idx = 1
+
+        # Print information
+        self._print_info = print_info
+
+    def __repr__(self) -> str:
+
+        s = "Time-stepping control object with atributes:\n"
+        s += f"Initial simulation time = {self.time_init}\n"
+        s += f"Final simulation time = {self.time_final}\n"
+        s += f"Initial time step = {self.dt_init}\n"
+        s += f"Minimum time step = {self.dt_min}\n"
+        s += f"Maximum time step = {self.dt_max}\n"
+        s += f"Lower optimal iteration range = {self.iter_low}\n"
+        s += f"Upper optimal iteration range = {self.iter_upp}\n"
+        s += f"Lower multiplication factor = {self.iter_low_factor}\n"
+        s += f"Upper multiplication factor = {self.iter_upp_factor}\n"
+        s += f"Recompute solution multiplication factor = {self.recomp_factor}\n"
+        s += f"Maximum recomputing attempts = {self.recomp_max}"
+
```

Comment:
```suggestion
s += f"Current time step and time are {self.dt} and {self.time}."
```

Reply:
Excellent idea. I will add the suggested change.